IPC can call directly into Swift

adetaylor opened the pull request:
#### ecc38014aa111ed3ef7e47e73b4dc67fe512b0c3
<pre>
IPC can call directly into Swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=302950">https://bugs.webkit.org/show_bug.cgi?id=302950</a>
<a href="https://rdar.apple.com/165213339">rdar://165213339</a>

Reviewed by Richard Robinson.

This adds the ability for our CoreIPC message dispatch to call directly
into Swift handlers.

GOALS:

The goals of this approach are:
* We do not duplicate all the message decoding code. The existing C++
  autogenerated decoding code continues to be used; however the dispatch
  mechanism then passes its results into Swift handler
  functions, by means of Swift/C++ interop.
* Messages can be defined, as now, within .in files, and each message calls
  directly into Swift handler functions, with no need for manually-written
  intermediate C++ shim functions for each message.

The key part is there&apos;s no need for manually-written C++ code for EVERY
MESSAGE.

USER EXPERIENCE:

To make a Swift message handler, you must:

* Mark your messages.in with a new attribute, SwiftReceiver.
* Create a Swift class with suitable functions for each message. It does
  not need any particular inheritance or conformance to any particular
  protocol. It needs to be named the same as the type named in the
  messages.in, just as if you were using C++.
* Within that class, create a new message _forwarder_ class
  (this is explained below) using code like this:
     WebKit.IPCTesterReceiverSwiftMessageForwarder.create(target: self);
  (the implementation of that &apos;create&apos; function is autogenerated; you
  don&apos;t have to do that)
* Register _that_ object to receive messages, rather than the Swift
  class itself.
* Keep that forwarder object referenced until you don&apos;t need to
  receive messages any more.

All of this is explained below, but users don&apos;t really need to know
anything more than the above.

An example is included in this PR - see IPCTesterReceiverSwift.

OBJECTS INVOLVED:

Somewhere, we have a Swift type, which actually handles the messsages.
Let&apos;s assume this is called IPCTesterReceiverSwift. This is a Swift class, thus
subject to reference counting. Note it&apos;s a pure Swift type, not some
hybrid C++/Swift type.

It contains methods which handle messages. Examples:

  @_expose(Cxx)
  internal func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {

We want the autogenerated code to call such methods.

However, our code generation currently generates C++ member functions
such as ::didReceiveMessage. As IPCTesterReceiverSwift is a Swift type, we can&apos;t
possibly attach C++ member functions to this type.

As such, we need an extra type. We&apos;re going to autogenerate this within
our Python code generation. We will call it
IPCTesterReceiverSwiftMessageForwarder.

It&apos;s a pure C++ type. It exists to accommodate the ::didReceiveMessage
(and similar) member functions. It does not have member functions for
each individual message; it&apos;s entirely autogenerated. It knows how to
forward actual individual method calls onto the real IPCTesterReceiverSwift (more
details on that later).

The code which registers IPCTesterReceiverSwift as a message receiver can no
longer register the IPCTesterReceiverSwift itself. Instead, it needs to
register the IPCTesterReceiverSwiftMessageForwarder.

OWNERSHIP:

The ownership here is a little fiddly.

The IPCTesterReceiverSwift OWNS the IPCTesterReceiverSwiftMessageForwarder, via
a Ref&lt;T&gt;. This is necessary because the IPC message infrastructure only retains
weak references to message receiver objects, so if it&apos;s not owned
elsewhere, it would be deallocated and no messages would be
received.

But, of course, the IPCTesterReceiverSwiftMessageForwarder needs to call
methods on the Swift IPCTesterReceiverSwift itself. To avoid a reference loop,
the IPCTesterReceiverSwiftMessageForwarder has a weak reference to
IPCTesterReceiverSwift.  Unfortunately, Swift/C++ interop does not yet have
built-in facilities for C++ to keep a weak Swift reference (this is
<a href="https://rdar.apple.com/165068975">rdar://165068975</a>) so an extra IPCTesterReceiverSwiftWeakRef class needs to be
created to fulfil this purpose. But it&apos;s essentially just a weak
reference.

When the IPCTesterReceiverSwiftMessageForwarder wants to call a
IPCTesterReceiverSwift method, it first calls
IPCTesterReceiverSwiftMessageForwarder::getTarget().
This will briefly turn the weak reference into a strong reference
while the method call occurs.

DISPATCH:

So, when C++ calls IPCTesterReceiverSwiftMessageForwarder::didReceiveMessage,
that decodes the message body in the normal way, and then
all this happens:
* didReceiveMessage calls IPCTesterReceiverSwiftMessageForwarder::getTarget
* that converts the weak reference briefly into a strong
  reference
* that strong reference is used to call the actual message
  dispatch function on the Swift IPCTesterReceiverSwift.
* Swift handles the message.

MESSAGES.IN CHANGES:

There&apos;s only one change to messages.in file, which is an optional
SwiftReceiver attribute. It&apos;s a Boolean. If that&apos;s enabled, all
the above stuff happens.

PASSING IPC CONNECTIONS:

Some message handlers like to receive an IPC::Connection. Currently,
they&apos;re passed as a C++ reference, which is not compatible with
Swift/C++ interop. So, if these functions are Swift, we need to pass
a pointer instead.

PASSING COMPLETION HANDLERS:

The pre-existing code passed completion handlers using C++ move
semantics. This is not yet supported by Swift/C++ interop
(<a href="https://rdar.apple.com/162361370">rdar://162361370</a>) so instead we have a wrap it in a
RefCountable wrapper.

RESTRICTIONS ON SWIFT TYPE USE:

The types in Swift can be forward-declared anywhere, but to actually
use them or know their size, we must include &lt;WebKit-Swift.h&gt;.
That may ONLY be done from .cpp files rather than .h files, to avoid
the risk of circular dependencies. (The generation of WebKit-Swift.h
depends on Swift being able to read potentially all of WebKit&apos;s header
files).

This constraint explains some of the delicate dance we do where
references to the Swift types are actually stored in std::unique_ptr;
this means the size of those types needs only be known by .cpp files
not by .h files. It&apos;s a bit of a pain and requires a few more heap
allocations, but for now, this is not performance-critical.

WHAT IS IN THIS COMMIT:

This commit contains all the above, but also a few other things
that merit explanation.

* More overloads of callMemberFunction which accept these two
  changes:
  * Use of a RefCountable&lt;CompletionHandler&gt; rather than CompletionHandler
  * Passing an IPC::Connection* rather than IPC::Connection&amp;
  (We may not need so many overloads).
* messages.py supports receivers outside namespaces since Swift
  code is not namespaced
* messages.py code to generate the IPCTesterReceiverSwiftMessageForwarder
  and IPCTesterReceiverSwiftWeakRef
* messages.py code to dispatch the handler function call to &apos;target&apos;
  rather than &apos;this&apos;, which is the same thing unless Swift is in use,
  in which case it&apos;s the thing returned by IPCTesterReceiverSwiftMessageForwarder::getTarget
* messages.py includes WebKit-Swift.h if the handler is in Swift code
  not C++
* a flag to enable us to expose &apos;internal&apos; visibility Swift things to C++

Canonical link: <a href="https://commits.webkit.org/303980@main">https://commits.webkit.org/303980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e99df947d9aebd74b7e40d75c2718e244993808d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134220 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6734 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141804 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d1924965-c917-4c0c-ace4-52b6a4733f0f) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7275 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6597 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102637 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/21c19764-cfb3-49bc-a49a-173085600bfe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137167 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120310 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83430 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/61ced47d-bc21-4b67-acd2-3a235bab0050) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/133569 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4952 "Build is in progress. Recent messages:") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/1616 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144447 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6404 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39029 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6488 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5389 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111265 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28208 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4808 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116582 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60179 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6456 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69922 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6409 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

adetaylor commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -148,6 +149,16 @@ void callMemberFunction(T* object, MF U::* function, ArgsTuple&& tuple, Completi
         }, std::forward<ArgsTuple>(tuple));
 }
 
+template<typename T, typename U, typename MF, typename ArgsTuple, typename CH>
```
All the new functions in this file are copy-and-paste of existing functions with either or both of these changes:
* Pass a pointer to the `IPC::Connection` rather than a reference
* Wrap the completionHandler in `WTF::RefCountable`.
The contents of the functions are the same as existing functions.

adetaylor commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -437,7 +437,7 @@ GENERATE_SINGLE_SWIFT_INTEROP_FILE_YES = GENERATE_SINGLE_SWIFT_INTEROP_FILE;
 // FIXME: (rdar://103177280) "-no-verify-emitted-module-interface;" is needed to workaround a SwiftVerifyEmittedModuleInterface bug
 OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS_SUPPORTS_SWIFT_OBJCXX_INTEROP_$(WK_SUPPORTS_SWIFT_OBJCXX_INTEROP)) $(OTHER_SWIFT_FLAGS_MARK_FOUNDATION_AS_OBJC_$(WK_GENERATE_SINGLE_SWIFT_INTEROP_FILE);
-OTHER_SWIFT_FLAGS_SUPPORTS_SWIFT_OBJCXX_INTEROP_YES = -Xcc -std=c++2b -no-verify-emitted-module-interface;
+OTHER_SWIFT_FLAGS_SUPPORTS_SWIFT_OBJCXX_INTEROP_YES = -Xcc -std=c++2b -no-verify-emitted-module-interface  -Xfrontend -emit-clang-header-min-access -Xfrontend internal;
```
Ignore this change - I'll be raising this as a separate PR as it's orthogonal to the use of Swift in IPC specifically (it is needed for us to use Swift anywhere in WebKit internals)

adetaylor replied:
That turned out to be impossible. It needs to be part of this change after all. I've conditionalized it on the enablement of this feature.

adetaylor commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -364,18 +480,34 @@ void handleMessageSynchronous(Connection& connection, Decoder& decoder, UniqueRe
-        callMemberFunction(object, function, connection, WTFMove(*arguments), WTFMove(completionHandler));
-    else
-        callMemberFunction(object, function, WTFMove(*arguments), WTFMove(completionHandler));
+        if constexpr (ValidationType::expectsConnectionArgument) {
```
The changes to this file aren't anywhere near as significant as they seem. It's just copypasta inside `if/else` conditions representing whether we're calling Swift or C++ receivers, with the slightly different ways they need to receive their arguments.

adetaylor commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -364,18 +480,34 @@ void handleMessageSynchronous(Connection& connection, Decoder& decoder, UniqueRe
-    CompletionHandlerType completionHandler(
+    logMessage(connection, MessageType::name(), object, *arguments);
+    if constexpr (ValidationType::completionHandlerIsCelled) {
+        Ref<CompletionHandlerType> completionHandler = CompletionHandlerType::create(
```
I may be able to remove some of the duplication here with additional template magic.

adetaylor replied:
I made a bit of progress at reducing the changes here, but there's more to do. Working on it, but the complexity here shouldn't be off-putting. The changes aren't anywhere near as significant as they look.

sheeparegreat commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -453,7 +453,7 @@
 #endif
 
-#if ENABLE(SWIFT_DEMO_URI_SCHEME)
+#if ENABLE(SWIFT_DEMO_URI_SCHEME) || ENABLE(IPC_TESTNG_SWIFT)
```
typo

sheeparegreat commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,55 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if ENABLE_IPC_TESTING_API
```
There is the bracket syntax for this, eg: ENABLE(IPC_TESTING_API) && ENABLE(IPC_TESTING_SWIFT)

adetaylor replied:
Yep, but it doesn't work in Swift

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwift.messages.in:
```diff
@@ -0,0 +1,30 @@
+# Copyright (C) 2021 Apple Inc. All rights reserved.
```
copyright years need updating in this file and others

adetaylor replied:
I've done this for the manually-written files.
If I update the copyright years in the `messages.py` code generator, then regenerate all our test files, it is a substantial code change, distracting from the real content of this PR, so I'll do that later shortly before submission.

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.cpp:
```diff
@@ -0,0 +1,129 @@
+        IPC::handleMessageAsync<Messages::TestWithSwift::TestAsyncMessage>(connection, decoder, target.get(), &TestWithSwift::testAsyncMessage);
+        return;
+    }
+    UNUSED_PARAM(connection);
```
if its unused can the parameter name just be omitted from the function signature?

adetaylor replied:
It's pre-existing autogenerated code, beyond the scope of this PR

adetaylor replied:
Pre-existing autogenerated code, beyond the scope of this PR

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift:
```diff
@@ -0,0 +1,51 @@
+internal import WebKit_Internal
+
+@_expose(Cxx)
+internal class TestWithSwiftWeakRef {
```
`internal` should be omitted since it’s the default

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @_expose(Cxx)
+    internal func getMessageTarget() -> TestWithSwift? {
```
ditto about internal

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift:
```diff
@@ -0,0 +1,51 @@
+        self.target = target
+    }
+
+    @_expose(Cxx)
```
is this needed? I thought everything is exposed to Cxx by default?

adetaylor replied:
Good question. I'm confused about this too. I'll try to work it out.

adetaylor replied:
Yeah you're right. Quite surprised about that. I assumed when we used the special flag to allow internal-level exports from Swift to C++, we'd need to be a bit specific to avoid _gajillions_ of things ending up in the interop header file. Apparently not. Presumably we have _gajillions_ of things in it. Seems to work though!

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.cpp:
```diff
@@ -0,0 +1,129 @@
+#include "config.h"
+#include "Shared/WebKit-Swift.h"
+
+#include "Decoder.h" // NOLINT
```
whats up with the NOLINT stuff?

adetaylor replied:
It's pre-existing autogenerated code, beyond the scope of this PR

adetaylor replied:
Pre-existing autogenerated code, beyond the scope of this PR

rr-codes commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1821,6 +1933,46 @@ def collect_message_statements(messages, message_statement_function):
     return ''.join(result)
 
 
+def generate_swift_message_handler(receiver):
```
we should really, really, really, really use Swift Macros for this, it would give such a better experience and functionality ( know that’s not practical in the short term, but just putting it out there!)

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,56 @@
+// Proxy interface to test IPC activities related to receiving messages in Swift.
+@_expose(Cxx)
+internal class IPCTesterReceiverSwift {
+    // Optional just because of an initialization order issue. Always occupied after initialization finished.
```
do you have more info/details about this?

adetaylor replied:
It's because we need to pass `self` into the initialiser for this data member.

rr-codes replied:
This pattern should almost always be avoided; would it be possible to do something like

```
self.messageForwarder = WebKit.IPCTesterReceiverSwiftMessageForwarder.create()
self.messageForwarder.target = self
```

?

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @_expose(Cxx)
+    internal func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {
```
we need to figure out how to allow async and/or throwing functions here (not urgent, but at some future point)

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,56 @@
+
+    @_expose(Cxx)
+    internal func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {
+        completionHandler.pointee(data + 1)
```
where is the `pointee` thing coming from?

adetaylor replied:
That is a built-in Swift/C++ interop thing. The `pointee` is the referent of `operator*`. And the `()` here is invocation of C++ `operator()`. This is not, I'm told, documented anywhere!

rr-codes replied:
this can probably be made nicer by defining an operator()() on RefCountable so then it won’t be needed to call .pointee directly

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,56 @@
+// THE POSSIBILITY OF SUCH DAMAGE.
+
+#if ENABLE_IPC_TESTING_API
+#if ENABLE_IPC_TESTING_SWIFT
```
this can just be &&’d with the previous one right?

adetaylor replied:
Oh perhaps, I'll try it

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,56 @@
+
+// Proxy interface to test IPC activities related to receiving messages in Swift.
+@_expose(Cxx)
+internal class IPCTesterReceiverSwift {
```
I have ideas for how there can be less boiler plate required by creating a protocol, I’ll describe so in a bit

adetaylor commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -384,6 +384,9 @@ WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR_cocoatouch = $(PRIVATE_HEADERS_FOLDER_PATH
 EXCLUDED_MACOS_PLUGIN_FILE_NAMES[sdk=iphone*] = SecItemShim.dylib WebProcessShim.dylib *.pdf Resources/mac/* com.apple.WebKit.NetworkProcess.sb com.apple.WebKit.GPUProcess.sb com.apple.WebKit.webpushd.mac.sb com.apple.WebProcess.sb;
 
+SWIFT_VERSION = 6.0;
+SWIFT_VERSION[sdk=macosx14*] = 5.0;
```
These two lines might not be submitted as part of this PR. I'm trying to use them to diagnose some EWS failures. I'd discuss with Elliott before landing them.

rr-codes replied:
this shouldn’t be needed because there exists

```
SWIFT_VERSION = $(SWIFT_VERSION$(WK_XCODE_16));
SWIFT_VERSION_XCODE_BEFORE_16 = 5.0;
SWIFT_VERSION_XCODE_SINCE_16 = 6.0;
```

in Base.xcconfig in WebKit

adetaylor replied:
That's right! But those lines aren't working. Without the extra lines here, we get EWS failures because the files are being interpreted at Swift 5.0 even if there is compiler version 6.2. Apparently. That's what I'm discussing with @emw-apple because I don't understand why.

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift:
```diff
@@ -0,0 +1,58 @@
+
+internal import WebKit_Internal
+
+class TestWithSwiftWeakRef {
```
all Swift classes should be `final` usually

adetaylor replied:
Done

rr-codes commented on Source/WebKit/Scripts/webkit/tests/TestWithSwiftMessageReceiver.swift:
```diff
@@ -0,0 +1,58 @@
+internal import WebKit_Internal
+
+class TestWithSwiftWeakRef {
+    weak var target: TestWithSwift?
```
can this be private?

adetaylor replied:
Done

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    func getMessageReceiver() -> RefIPCTesterReceiverSwiftMessageForwarder {
+        guard let messageForwarder = self.messageForwarder else {
```
and then all this can be avoided if the above is able to be resolved since then this can be non-nil at compile time

adetaylor replied:
Yeah, unfortunately (as discussed in a chat thread) there's really no way to avoid this. Swift is diligent about making sure everything is initialised before allowing you to pass `self` around, which is good!

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,50 @@
+        return messageForwarder
+    }
+
+    func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {
```
can this be named something more realistic, like `computeSuccessiveValue` or something like that

adetaylor replied:
I'm trying to match the names and functionality of the pre-existing identical C++ tester, so I'll keep it as-is.

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwift.swift:
```diff
@@ -0,0 +1,50 @@
+        return messageForwarder
+    }
+
+    func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {
```
in general, Swift prefers types like `Int` over `UInt32` unless there is a specific reason why a specific width integer type is needed

adetaylor replied:
I'm trying to match the names and functionality of the pre-existing identical C++ tester, so I'll keep it as-is.

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwiftTypes.h:
```diff
@@ -25,30 +25,8 @@
-} // namespace WebCore
-
-#endif // ENABLE(WEBXR_HIT_TEST)
+using IntCompletionHandler = WTF::RefCountable<WTF::CompletionHandler<void(uint32_t)>>;
```
I think this file should be named like “IPCTesterReceiverTypes” or “IPCTesterReceiverBridging” since it’s a pet peeve of mine when a file name has a language name in it, and also if it has “Utilities” in it 😛

adetaylor replied:
I've renamed it to `IPCTesterReceiverSwiftTypes`. I agree 'utilities' is no good.
However, `IPCTesterReceiverSwift` is the name of the test message handler, as distinct from the pre-existing `IPCTesterReceiver` which receives and handles similar messages in C++. I could `s/Swift/Two/g` or something. I'll think about that.

rr-codes commented on Source/WebKit/Shared/IPCTesterReceiverSwiftTypes.h:
```diff
@@ -25,30 +25,8 @@
-} // namespace WebCore
-
-#endif // ENABLE(WEBXR_HIT_TEST)
+using IntCompletionHandler = WTF::RefCountable<WTF::CompletionHandler<void(uint32_t)>>;
```
I think this type should be named like `{function name}CompletionHandler` so like `AsyncMessageCompletionHandler`

adetaylor replied:
Done. I'll also look into autogenerating them in a follow-up change.

rr-codes reviewed (approved):
this is good for now, but please file a bug to track improving the Swift ergonomics (I’ll message you with some ideas I have to do this). Also might want to get @achristensen07 or @aprotyas ’s look at some of the fancy C++ templates stuff

adetaylor committed:
IPC can call directly into Swift
https://bugs.webkit.org/show_bug.cgi?id=302950
rdar://165213339

Reviewed by Richard Robinson.

This adds the ability for our CoreIPC message dispatch to call directly
into Swift handlers.

GOALS:

The goals of this approach are:
* We do not duplicate all the message decoding code. The existing C++
  autogenerated decoding code continues to be used; however the dispatch
  mechanism then passes its results into Swift handler
  functions, by means of Swift/C++ interop.
* Messages can be defined, as now, within .in files, and each message calls
  directly into Swift handler functions, with no need for manually-written
  intermediate C++ shim functions for each message.

The key part is there's no need for manually-written C++ code for EVERY
MESSAGE.

USER EXPERIENCE:

To make a Swift message handler, you must:

* Mark your messages.in with a new attribute, SwiftReceiver.
* Create a Swift class with suitable functions for each message. It does
  not need any particular inheritance or conformance to any particular
  protocol. It needs to be named the same as the type named in the
  messages.in, just as if you were using C++.
* Within that class, create a new message _forwarder_ class
  (this is explained below) using code like this:
     WebKit.IPCTesterReceiverSwiftMessageForwarder.create(target: self);
  (the implementation of that 'create' function is autogenerated; you
  don't have to do that)
* Register _that_ object to receive messages, rather than the Swift
  class itself.
* Keep that forwarder object referenced until you don't need to
  receive messages any more.

All of this is explained below, but users don't really need to know
anything more than the above.

An example is included in this PR - see IPCTesterReceiverSwift.

OBJECTS INVOLVED:

Somewhere, we have a Swift type, which actually handles the messsages.
Let's assume this is called IPCTesterReceiverSwift. This is a Swift class, thus
subject to reference counting. Note it's a pure Swift type, not some
hybrid C++/Swift type.

It contains methods which handle messages. Examples:

  @_expose(Cxx)
  internal func asyncMessage(data: UInt32, completionHandler: IntCompletionHandler) {

We want the autogenerated code to call such methods.

However, our code generation currently generates C++ member functions
such as ::didReceiveMessage. As IPCTesterReceiverSwift is a Swift type, we can't
possibly attach C++ member functions to this type.

As such, we need an extra type. We're going to autogenerate this within
our Python code generation. We will call it
IPCTesterReceiverSwiftMessageForwarder.

It's a pure C++ type. It exists to accommodate the ::didReceiveMessage
(and similar) member functions. It does not have member functions for
each individual message; it's entirely autogenerated. It knows how to
forward actual individual method calls onto the real IPCTesterReceiverSwift (more
details on that later).

The code which registers IPCTesterReceiverSwift as a message receiver can no
longer register the IPCTesterReceiverSwift itself. Instead, it needs to
register the IPCTesterReceiverSwiftMessageForwarder.

OWNERSHIP:

The ownership here is a little fiddly.

The IPCTesterReceiverSwift OWNS the IPCTesterReceiverSwiftMessageForwarder, via
a Ref<T>. This is necessary because the IPC message infrastructure only retains
weak references to message receiver objects, so if it's not owned
elsewhere, it would be deallocated and no messages would be
received.

But, of course, the IPCTesterReceiverSwiftMessageForwarder needs to call
methods on the Swift IPCTesterReceiverSwift itself. To avoid a reference loop,
the IPCTesterReceiverSwiftMessageForwarder has a weak reference to
IPCTesterReceiverSwift.  Unfortunately, Swift/C++ interop does not yet have
built-in facilities for C++ to keep a weak Swift reference (this is
rdar://165068975) so an extra IPCTesterReceiverSwiftWeakRef class needs to be
created to fulfil this purpose. But it's essentially just a weak
reference.

When the IPCTesterReceiverSwiftMessageForwarder wants to call a
IPCTesterReceiverSwift method, it first calls
IPCTesterReceiverSwiftMessageForwarder::getTarget().
This will briefly turn the weak reference into a strong reference
while the method call occurs.

DISPATCH:

So, when C++ calls IPCTesterReceiverSwiftMessageForwarder::didReceiveMessage,
that decodes the message body in the normal way, and then
all this happens:
* didReceiveMessage calls IPCTesterReceiverSwiftMessageForwarder::getTarget
* that converts the weak reference briefly into a strong
  reference
* that strong reference is used to call the actual message
  dispatch function on the Swift IPCTesterReceiverSwift.
* Swift handles the message.

MESSAGES.IN CHANGES:

There's only one change to messages.in file, which is an optional
SwiftReceiver attribute. It's a Boolean. If that's enabled, all
the above stuff happens.

PASSING IPC CONNECTIONS:

Some message handlers like to receive an IPC::Connection. Currently,
they're passed as a C++ reference, which is not compatible with
Swift/C++ interop. So, if these functions are Swift, we need to pass
a pointer instead.

PASSING COMPLETION HANDLERS:

The pre-existing code passed completion handlers using C++ move
semantics. This is not yet supported by Swift/C++ interop
(rdar://162361370) so instead we have a wrap it in a
RefCountable wrapper.

RESTRICTIONS ON SWIFT TYPE USE:

The types in Swift can be forward-declared anywhere, but to actually
use them or know their size, we must include <WebKit-Swift.h>.
That may ONLY be done from .cpp files rather than .h files, to avoid
the risk of circular dependencies. (The generation of WebKit-Swift.h
depends on Swift being able to read potentially all of WebKit's header
files).

This constraint explains some of the delicate dance we do where
references to the Swift types are actually stored in std::unique_ptr;
this means the size of those types needs only be known by .cpp files
not by .h files. It's a bit of a pain and requires a few more heap
allocations, but for now, this is not performance-critical.

WHAT IS IN THIS COMMIT:

This commit contains all the above, but also a few other things
that merit explanation.

* More overloads of callMemberFunction which accept these two
  changes:
  * Use of a RefCountable<CompletionHandler> rather than CompletionHandler
  * Passing an IPC::Connection* rather than IPC::Connection&
  (We may not need so many overloads).
* messages.py supports receivers outside namespaces since Swift
  code is not namespaced
* messages.py code to generate the IPCTesterReceiverSwiftMessageForwarder
  and IPCTesterReceiverSwiftWeakRef
* messages.py code to dispatch the handler function call to 'target'
  rather than 'this', which is the same thing unless Swift is in use,
  in which case it's the thing returned by IPCTesterReceiverSwiftMessageForwarder::getTarget
* messages.py includes WebKit-Swift.h if the handler is in Swift code
  not C++
* a flag to enable us to expose 'internal' visibility Swift things to C++

Canonical link: https://commits.webkit.org/303980@main

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -361,7 +361,13 @@ INCLUDED_EXTENSION_SOURCE_FILE_NAMES[sdk=iphone*17.2.internal] = ;
-EXCLUDED_SOURCE_FILE_NAMES = $(EXCLUDED_PRODUCT_DEPENDENCY_NAMES) $(EXCLUDED_IOS_RESOURCE_FILE_NAMES) $(EXCLUDED_MACOS_PLUGIN_FILE_NAMES) $(EXCLUDED_MIGRATED_HEADERS_COCOA_TOUCH_$(WK_IS_COCOA_TOUCH)) $(WK_EXCLUDED_WEBPUSHD_SANDBOX) $(EXCLUDED_EXTENSION_SOURCE_FILE_NAMES) $(EXCLUDED_SWIFT_FILE_NAMES) $(WK_EXCLUDED_SEPARATED_IMAGE_FILES) $(WK_EXCLUDED_SWIFTOVERLAY_SOURCE_FILE_NAMES);
+EXCLUDED_SOURCE_FILE_NAMES = $(EXCLUDED_PRODUCT_DEPENDENCY_NAMES) $(EXCLUDED_IOS_RESOURCE_FILE_NAMES) $(EXCLUDED_MACOS_PLUGIN_FILE_NAMES) $(EXCLUDED_MIGRATED_HEADERS_COCOA_TOUCH_$(WK_IS_COCOA_TOUCH)) $(WK_EXCLUDED_WEBPUSHD_SANDBOX) $(EXCLUDED_EXTENSION_SOURCE_FILE_NAMES) $(EXCLUDED_SWIFT_FILE_NAMES) $(WK_EXCLUDED_SEPARATED_IMAGE_FILES) $(WK_EXCLUDED_SWIFTOVERLAY_SOURCE_FILE_NAMES) $(WK_EXCLUDED_S ...
+
+// Older SDKs can't parse this file since it has the `unsafe` keyword,
+// and that causes problems even if it's inside #if SOMETHING_THAT_IS_NOT_ENABLED
+WK_EXCLUDED_SWIFT_IPC_FILE_NAMES = $(WK_EXCLUDED_SWIFT_IPC_FILE_NAMES_$(ENABLE_IPC_TESTING_SWIFT));
+WK_EXCLUDED_SWIFT_IPC_FILE_NAMES_ = IPCTesterReceiverSwiftMessageReceiver.swift;
+WK_EXCLUDED_SWIFT_IPC_FILE_NAMES_ENABLE_IPC_TESTING_SWIFT = ;
```
For posterity, I think this is due to https://github.com/swiftlang/swift/pull/74415 interacting with the way the `#if compiler` check is nested in the messages.py generator. This fix did not ship until Swift 6.1, and our build config for Sonoma uses 5.10.

adetaylor replied:
Attempting to use this wisdom in https://github.com/WebKit/WebKit/pull/55092.
